Block comparison: show difference between invalid block conversion options

johngodley opened the pull request:
## Description

When you have an invalid block the result of the available actions can be unknown.

![edit_post_ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/44956216-58184080-aeb8-11e8-903f-b148ef1a1af6.jpg)

This PR adds a 'compare conversion' option that shows the actual HTML difference when 'convert to blocks' is applied.

<img src="https://user-images.githubusercontent.com/1277682/42805400-9d80ad2c-89a3-11e8-97cf-36e72c069417.jpg" width="400" />

The difference appears in a modal and the user can select to keep the current contents or apply the conversion, with full knowledge of the result of their action.

![another_post___ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/42805227-2d622be2-89a3-11e8-929f-ee2fb78e6270.jpg)

The modal adjusts to the height of the block:

![edit_post_ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/44956332-18525880-aeba-11e8-9b3c-2b38e28a0d4d.jpg)

Overflowing, if too tall, and scrollable:

![edit_post_ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/44956339-3b7d0800-aeba-11e8-8625-9e85d1b12cd2.jpg)

And adjusts accordingly on smaller screens (using overflow scroll if too tall):

![edit_post_ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/44956345-518ac880-aeba-11e8-86c0-cea06c0f66ee.jpg)

Note: this is based on the discussion in #7604 

## How has this been tested?

1. Manually check that the 'compare conversion' option displays correctly for different types of blocks and at different screen sizes
2. Manually verify that the 'keep as HTML' converts the block to an HTML block
3. Manually verify that the 'convert to blocks' applies a conversion identical to picking 'convert to blocks' from the invalid block modal

## Types of changes
This adds a new feature.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

johngodley committed:
Add block compare component

Shows two blocks side by side with the difference in HTML highlighted

johngodley committed:
Block wanring: add comparison option

Use the block comparison component to show the difference between the invalid block options in a modal

johngodley commented:
Noting that the comparison currently uses code from `jsdiff` to perform proper diffing. This is a BSD licensed open source library. It works well, but we don't need the entire library. I'm unsure about the best route for proceeding here, and I don't know if it's appropriate to include the library as-is, or the full package.

While Gutenberg does have it's own HTML parsing, and also HTML comparison (via `isEquivalentHTML`), it's doesn't find the longest common sequences, and would result in a poor diff that would need adapting to a proper [diffing algorithm](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.4.6927).

An alternative is to re-implement `jsdiff`  ourselves.

Open to suggestions.

johngodley commented:
Also noting that the style is a best attempt at matching that proposed in #7604, but needs some improving. @jasmussen I'd very much appreciate any suggestions/help you can give.

In particular I assumed that we would want to show deleted characters in the right panel, and I've included these in red. We can not show those at all, if it makes more sense.

Commit:
Small fix.

jasmussen commented:
@johngodley This is impressive work. I think it's super duper solid, and it's VERY much good enough to get into master so we can iterate. 

I pushed a small fix so it only scrolls when necessary, and I made the menu option titlecase. 

Can you rebase this? 

Question: does this use the modal API? Assuming yes as it really seems to be, which is great, just verifying.

johngodley commented:
> does this use the modal API

If you mean `<Modal />` then yep it does.

Rebased. There's a codecov failure which I don't understand as the unit test coverage has increased.

I've noticed some problems with a few of the more esoteric blocks so I'll look at fixing those.

johngodley committed:
Fix line heights for deleted content

If a deleted content span went over onto the next line it would overlap with other spans

jasmussen commented:
> If you mean <Modal /> then yep it does.

Yes, awesome!

I'm extending the review range and adding a milestone so we can get this in. Amazing work! This is a KILLER feature.

johngodley commented:
I made a some small tweaks to the CSS regarding line height. If you had a deleted section wrapping over lines it could overlap:

![edit_post_ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/45147830-c3feef80-b1bd-11e8-8f53-43bd36c2c724.jpg)

This has now been fixed.

I noticed two other problems:
- The `Modal` component doesn't prevent focus/blur events from propagating through to the page underneath. This means you can trigger tooltips and block dropdown menus, making it a bit twitchy feeling. I've got a fix in place which I'll put in a separate PR
- Some blocks don't behave well to 'convert to blocks', and can result in the entire block being deleted. This isn't caused by the code here, but the preview ability makes it more noticeable. I'll also look at this separately

jasmussen commented:
> The Modal component doesn't prevent focus/blur events from propagating through to the page underneath. This means you can trigger tooltips and block dropdown menus, making it a bit twitchy feeling. I've got a fix in place which I'll put in a separate PR

🏅 

Yeah this is a separate issue that would be sweet to fix.

> Some blocks don't behave well to 'convert to blocks', and can result in the entire block being deleted. This isn't caused by the code here, but the preview ability makes it more noticeable. I'll also look at this separately

You rule the world.

johngodley committed:
Fix flexbox spacing when converted preview is smaller

The vertical spacing would become strange if the converted preview was shorter than the original preview

johngodley committed:
Update snapshot for flexbox fixes

talldan reviewed (changes requested):
@johngodley - This is really cool. In fact, I think it's so useful I was wondering if we would want to make it more prominent?

I added some comments, I think the main one that needs to be fixed is the `max-height: 70%` issue which is visible to the user.

Will probably also need someone more knowledgeable to comment on the library inclusion. Not sure who would be best.

johngodley committed:
Clean up var naming and class usage

johngodley committed:
Remove unnecessary flexbox

johngodley commented on packages/editor/src/components/block-compare/index.js:
```diff
@@ -0,0 +1,84 @@
+		const difference = differ.diff( originalContent, newContent );
+
+		return difference.map( ( item, pos ) => {
+			if ( item.added ) {
```
Excellent idea!

johngodley commented on packages/editor/src/components/block-compare/style.scss:
```diff
@@ -0,0 +1,75 @@
+	padding-bottom: $panel-padding;
+
+	> div {
+		display: flex;
```
It's being used to push the buttons to the same position at the bottom of the modal, otherwise they become vertically misaligned.

johngodley commented on packages/editor/src/components/block-list/block-invalid-warning.js:
```diff
@@ -14,50 +15,97 @@ import { withDispatch } from '@wordpress/data';
+
+		if ( compare ) {
+			return (
+				<Modal
```
Yeah, it's a good point. My thinking is that the comparison component could (potentially, and currently without any need) be used elsewhere that might be outside of a modal or an on/off situation and so I made the component do it's thing, and left it up to the parent to decide when to show it.

johngodley committed:
Remove max height

It’s not necessary for anything and makes the modal too small on mobile

johngodley committed:
Allow full height on small screens

johngodley commented:
> I think it's so useful I was wondering if we would want to make it more prominent?

Thanks for the review @talldan. I do wonder if it might make more sense if the 'convert to blocks' button just defaults to showing the comparison component directly, and we remove the option from the dropdown menu?

The effect of 'convert to blocks' is unknown to the user, and sometimes quite arbitrary and destructive. At the point the user is being presented with this warning I don't think it's excessive to add an additional click in the process. It saves having to undo the action and being annoyed at Gutenberg if the conversion isn't as expected.

johngodley committed:
Restore flex for vertical spacing of elements

Without this the button doesnt stick to the bottom

talldan commented on packages/editor/src/components/block-compare/index.js:
```diff
@@ -0,0 +1,89 @@
+/**
+ * External dependencies
+ */
+import classname from 'classnames';
```
Just a typo here, the imported function should be `classnames` instead of `classname`

talldan commented:
@johngodley This is looking good. I've spotted one additional thing here:
https://github.com/WordPress/gutenberg/pull/7995/files#r216273361

I think the only thing we need now is to confirm that the script/dependency inclusion is ok.

I feel like it could definitely be iterated on this so that it's part of the workflow when editing a block HTML:
1. User edits block HTML
2. The diff is present to them when they've finished editing with the two options.
3. User decides an option.

There are other times when the user might end up with invalid html (e.g. using the classic editor), so we might present that slightly differently to what I've proposed above. Would be good to get additional input on this though.

johngodley committed:
Replace inline diff code with jsdiff module

johngodley commented:
Thanks for that link, and it's pushed me towards just bundling the module normally

johngodley commented on packages/editor/src/components/block-compare/block-view.js:
```diff
@@ -0,0 +1,28 @@
+/**
```
I'm not sure what you mean by editors - can you point me in the direction?

jasmussen replied:
Can't speak for Andrew but since he's asleep I'll take an educated guess. Maybe he meant `edit-post`?

<img width="407" alt="screen shot 2018-09-12 at 10 49 29" src="https://user-images.githubusercontent.com/1204802/45413479-98767c00-b679-11e8-93f0-615c470a10eb.png">

jasmussen replied:
Again I can't but make educated guesses — I think it's something about the packages getting published on NPM but not sure.

talldan replied:
I'll wait for Andrew to clarify, but the component is already in the `editor` package, not in the `components` package, so maybe it was a misread.

aduth replied:
Oof, I'm sorry for the trouble, it was definitely a misread on my part.

I had mistaken it as being part of `packages/components`.

talldan commented on packages/editor/src/components/block-compare/index.js:
```diff
@@ -0,0 +1,88 @@
+
+		return {
+			rawContent: block.originalContent,
+			renderedContent: getSaveElement( blockType, block.attributes ),
```
I've just given this another test, and I'm not sure the original content's preview displays quite what I was expecting. My steps:

1. Add some simple text to a paragraph block
2. Edit the html and add an inline style `style="font-size:25px;"`
3. View the comparison
4. Note that the original content preview doesn't show the font size at 25px, which is what I'd get if I selected 'Keep as HTML'.

To make the preview display correctly I tried using RawHTML like the HTML block does for its preview, which seemed to do the trick:
```
	getOriginalContent( block ) {
		return {
			rawContent: block.originalContent,
			renderedContent: (
				<RawHTML>{ block.originalContent }</RawHTML>
			),
		};
	}
```

johngodley replied:
Interesting!

RawHTML uses `dangerouslySetInnerHTML`, which opens up things like `<p onClick="alert(1)">fdsfsdfds fd sfds</p>`. I'm not sure if this is a problem in this context and what the behaviour is supposed to be so I'll have a poke around.

johngodley replied:
The 'preview' mode of the HTML block uses the `<Sandbox />` component to display arbitrary HTML, which neutralises my above example, and seems to be the way to go. The styling is different so I'll see if it can be made to fit in.

talldan replied:
Hmm, yep, thanks for looking into it. Lets go with what we've got and then maybe look into this on a separate issue.

talldan reviewed (commented):
This is really close, I just spotted one thing in further testing. Will also wait for @aduth to clarify his comment.

johngodley commented:
Thanks for the feedback and suggestions!

jasmussen commented:
Amazing work!

aduth commented on packages/editor/src/components/block-compare/block-view.js:
```diff
@@ -0,0 +1,28 @@
+	return (
+		<div className={ className }>
+			<div className="editor-block-compare__content">
+				<h1 className="components-modal__header-heading">{ title }</h1>
```
This is not the modal component, so should not be stealing / inheriting styles. Components should be insulated from one another to avoid maintenance issues with cascading consequences of future changes.

aduth commented on packages/editor/src/components/block-compare/block-view.js:
```diff
@@ -0,0 +1,28 @@
+			</div>
+
+			<div className="editor-block-compare__action">
+				<Button isLarge tabindex="0" onClick={ action }>{ actionText }</Button>
```
I'm guessing a console warning is being logged on this line?

We should pass `tabIndex` as the DOM property `tabIndex={ 0 }`, not the attribute name.

>Note that you should still use the canonical React naming for known attributes:
>
>```
>// Yes, please
><div tabIndex="-1" />
>
>// Warning: Invalid DOM property `tabindex`. Did you mean `tabIndex`?
><div tabindex="-1" />
>```
>
>In other words, the way you use DOM components in React hasn’t changed, but now you have some new capabilities.

https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html

johngodley replied:
Oddly it doesn't log a warning, but both very valid points and I'll get those fixed in another PR
